File: Libro de Flutter/01-fundamentos-dart/01_hola_mundo.dart
// =============================================================================
// ARCHIVO: 01_hola_mundo.dart
// TEMA: Tu primer programa en Dart
// EJECUCION: dart run 01_hola_mundo.dart
// =============================================================================
//
// Este archivo es el punto de partida para entender Dart como lenguaje.
// Antes de construir interfaces en Flutter, necesitas comprender como un
// programa lee instrucciones, crea datos en memoria y produce una salida.
//
// Cuando ejecutes este archivo:
// - Dart compilara o interpretara el codigo segun el entorno.
// - El programa entrara por main().
// - Se crearan variables en memoria.
// - La consola mostrara la salida visible.
// =============================================================================

// ─────────────────────────────────────────────────────────────────────────────
// EL PUNTO DE ENTRADA: void main()
// ─────────────────────────────────────────────────────────────────────────────

// En Dart, todo programa comienza en la funcion main().
// Es el punto de entrada. Sin main(), el runtime no sabe que ejecutar primero.
// "void" significa que esta funcion no devuelve un valor al finalizar.
void main() {
  // ─────────────────────────────────────────────────────────────────────────
  // HOLA MUNDO — Lo más básico
  // ─────────────────────────────────────────────────────────────────────────

  // print() envia informacion a la consola.
  // Esa consola es una salida visible del software para mostrar resultados.
  print('Hola, mundo!');
  print('Bienvenido al Libro de Flutter.');

  // ─────────────────────────────────────────────────────────────────────────
  // IMPRIMIR DISTINTOS TIPOS DE VALORES
  // ─────────────────────────────────────────────────────────────────────────

  // Dart puede convertir muchos valores a texto automaticamente para imprimirlos.
  print(42);        // Número entero
  print(3.14);      // Número decimal
  print(true);      // Booleano
  print(null);      // El valor nulo (ausencia de valor)

  // ─────────────────────────────────────────────────────────────────────────
  // VARIABLES BÁSICAS
  // ─────────────────────────────────────────────────────────────────────────

  // "var" le pide a Dart inferir el tipo automaticamente.
  // El valor vive en memoria y la variable actua como un nombre para accederlo.
  var nombre = 'Flutter';
  var version = 3;
  var estaBuenisimo = true;

  print(nombre);
  print(version);
  print(estaBuenisimo);

  // ─────────────────────────────────────────────────────────────────────────
  // INTERPOLACIÓN DE STRINGS — El equivalente a los f-strings de Python
  // ─────────────────────────────────────────────────────────────────────────

  // La interpolacion inserta valores dentro del texto.
  // En lugar de unir cadenas manualmente, dejas que Dart construya el resultado.
  print('Hola desde $nombre $version!');

  // Para expresiones mas complejas, usa llaves: ${expresion}
  var precio = 9.99;
  var cantidad = 3;
  print('Total: \$${precio * cantidad}');
  // El simbolo \ evita que el primer $ se interprete como interpolacion.

  // Tambien puedes ejecutar operaciones dentro de la interpolacion.
  print('Nombre en mayúsculas: ${nombre.toUpperCase()}');
  print('Longitud del nombre: ${nombre.length} letras');

  // ─────────────────────────────────────────────────────────────────────────
  // STRINGS CON COMILLAS SIMPLES Y DOBLES
  // ─────────────────────────────────────────────────────────────────────────

  // Dart acepta comillas simples y dobles.
  // Lo importante es ser consistente para mejorar legibilidad.
  var mensaje1 = 'Usando comillas simples';
  var mensaje2 = "Usando comillas dobles";
  print(mensaje1);
  print(mensaje2);

  // Para strings multilínea, usa triple comilla (''' o """):
  var textoLargo = '''
Este es un string
que ocupa varias
líneas.
  ''';
  print(textoLargo);

  // ─────────────────────────────────────────────────────────────────────────
  // CONCATENACIÓN vs INTERPOLACIÓN
  // ─────────────────────────────────────────────────────────────────────────

  // Puedes concatenar strings con +, pero la interpolacion suele ser mas clara.
  var lenguaje = 'Dart';

  // Forma menos recomendada (concatenación):
  print('Aprendiendo ' + lenguaje + ' desde cero');

  // Forma recomendada (interpolación):
  print('Aprendiendo $lenguaje desde cero');
}

// =============================================================================
// QUE DEBERIAS ENTENDER AL TERMINAR
// - main() es el punto de entrada del programa.
// - print() produce una salida visible en consola.
// - Las variables apuntan a valores administrados por Dart en memoria.
// - La interpolacion vuelve mas claro el texto dinamico.
// - Dart distingue tipos y eso ayuda a evitar errores.
//
// PRACTICA GUIADA
// 1. Cambia el valor de nombre y version.
// 2. Crea una variable llamada ciudad.
// 3. Imprime un mensaje con tu nombre, ciudad y edad.
// 4. Intenta hacer print(nombre + version) y explica el error.
// 5. Calcula con ${} cuantos dias hay en 4 semanas.
// =============================================================================

File: Libro de JavaScript/01-fundamentos/01_hola_mundo.js
// =============================================================================
// ARCHIVO: 01_hola_mundo.js
// TEMA: Tu primer programa en JavaScript
// EJECUCION: node 01_hola_mundo.js
// =============================================================================
//
// Esta leccion introduce la idea mas importante de toda programacion:
// escribir instrucciones que otra pieza de software pueda ejecutar.
//
// En este caso, Node.js carga tu archivo, el motor de JavaScript interpreta
// el codigo, reserva memoria para valores y muestra resultados en consola.
// =============================================================================

// ─────────────────────────────────────────────────────────────────────────────
// 1. console.log() — La función más usada en JavaScript
// ─────────────────────────────────────────────────────────────────────────────

// console.log() envia informacion a la salida estandar.
// Dicho simple: le dice al programa que muestre algo en la consola.
console.log('Hola, mundo!');
console.log('Bienvenido a JavaScript');

// Puedes imprimir varios valores separados por comas en una sola llamada.
// JavaScript los separará automáticamente con un espacio.
console.log('Mi nombre es', 'Juan', 'y tengo', 25, 'años');

// Tambien puedes imprimir numeros, booleanos y otros tipos de dato.
console.log(42);
console.log(true);
console.log(3.14);

// ─────────────────────────────────────────────────────────────────────────────
// 2. Otros métodos de console
// ─────────────────────────────────────────────────────────────────────────────

// console.error() usa la salida de error.
// Muchos entornos la muestran diferente porque representa problemas o alertas.
console.error('Esto es un mensaje de error');

// console.warn() comunica una advertencia.
console.warn('Esto es una advertencia');

// console.table() muestra datos en formato de tabla — muy útil para arrays
// y objetos. Hace la información mucho más fácil de leer.
const personas = [
  { nombre: 'Ana', edad: 30 },
  { nombre: 'Luis', edad: 25 },
  { nombre: 'María', edad: 35 },
];
console.log('\n--- console.table() ---');
console.table(personas);

// ─────────────────────────────────────────────────────────────────────────────
// 3. Template literals — Strings modernos con backticks
// ─────────────────────────────────────────────────────────────────────────────

// Los template literals usan backticks (`).
// Su ventaja principal es la interpolacion: insertar variables o expresiones
// dentro del texto sin concatenar manualmente.

const nombre = 'Carlos';
const edad = 28;

// Forma antigua: unir texto y valores con +.
const mensajeAntiguo = 'Hola, ' + nombre + '! Tienes ' + edad + ' años.';
console.log(mensajeAntiguo);

// Forma moderna: un template literal.
const mensajeModerno = `Hola, ${nombre}! Tienes ${edad} años.`;
console.log(mensajeModerno);

// Dentro de ${} puedes poner cualquier expresión JavaScript válida.
const precio = 100;
const descuento = 0.15;
console.log(`Precio final: $${precio * (1 - descuento)}`);
console.log(`2 + 2 = ${2 + 2}`);
console.log(`Mayúsculas: ${nombre.toUpperCase()}`);

// ─────────────────────────────────────────────────────────────────────────────
// 4. Comentarios en JavaScript
// ─────────────────────────────────────────────────────────────────────────────

// Comentario de una sola línea: usa dos barras diagonales //
// Todo lo que venga después en esa línea es ignorado por JavaScript.

/*
  Comentario de múltiples líneas.
  Útil para explicaciones largas, documentación
  o para desactivar bloques de código temporalmente.
*/

// Buenas practicas:
// - comenta la intencion y el contexto;
// - no repitas exactamente lo que ya dice el codigo;
// - actualiza el comentario si el codigo cambia.

// ─────────────────────────────────────────────────────────────────────────────
// 5. Ejemplo completo
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Genera un saludo personalizado para un usuario.
 * Sirve para mostrar que una funcion puede recibir datos de entrada,
 * procesarlos y devolver una salida.
 * @param {string} nombreUsuario - Nombre de la persona a saludar.
 * @param {string} ciudad - Ciudad de origen de la persona.
 * @returns {string} Mensaje listo para imprimirse.
 */
const generarSaludo = (nombreUsuario, ciudad) => {
  return `¡Hola, ${nombreUsuario}! Bienvenido desde ${ciudad}.`;
};

console.log('\n--- Ejemplo completo ---');
console.log(generarSaludo('Sofía', 'Madrid'));
console.log(generarSaludo('Pedro', 'Ciudad de México'));

// =============================================================================
// QUE DEBERIAS ENTENDER AL TERMINAR
// - Node.js ejecuta tu archivo JavaScript fuera del navegador.
// - console.log() muestra informacion en la consola.
// - Las variables almacenan referencias a valores en memoria.
// - Un template literal hace mas clara la construccion de mensajes.
// - Una funcion recibe datos, hace algo y devuelve un resultado.
//
// PRACTICA GUIADA
// 1. Cambia el valor de nombre y edad.
// 2. Crea una variable ciudad y usala en un template literal.
// 3. Agrega un nuevo objeto al array personas.
// 4. Escribe una funcion que calcule el area de un rectangulo.
// 5. Explica con tus palabras la diferencia entre console.log y console.error.
// =============================================================================
